exercises/saddle-points/example.rs

```diff
@@ -1,23 +1,24 @@
 pub fn find_saddle_points(input: &[Vec<u64>]) -> Vec<(usize, usize)> {
-    let mut saddle_points = Vec::new();
+    let (width, height) = match (input.len(), input[0].len()) {
+        (0, _) | (_, 0) => return Vec::new(),
+        (n, k) => (n, k),
+    };
 
-    let width = input.len();
-    let height = input[0].len();
+    let mut saddle_points = Vec::new();
 
     for i in 0..width {
-        for j in 0..height {
-            let column = input.iter().map(|x| x[j]).collect::<Vec<u64>>();
-            let row = &input[i];
+        let row_max = input[i].iter().max().unwrap();
 
-            let max = row.iter().max().unwrap();
-            let min = column.iter().min().unwrap();
+        for j in 0..height {
+            let column_min = input.iter().map(|x| x[j]).min().unwrap();
```

Comment:
You're computing the column min for every `i` and `j`; computing that min is ~O(n), so overall, the column minima scan is O(n^3). It's more efficient to precompute a vector of column minima ahead of time, and then index into it; this can be done with O(n^2) time and O(n) memory, [like this](https://github.com/coriolinus/exercism-rs/blob/master/saddle-points/src/lib.rs):

```rust
            let col_min = (0..min_width)
                .map(|coln| input.iter().map(|row| row[coln]).min().unwrap())
                .collect::<Vec<_>>();
```

This isn't a huge deal, but you do say that this is an optimization, so let's make it optimal.

Reply:
Yep, that's an improvement that I'm aware of. Re-calculating column minimums for each matrix element can't be efficient. I didn't include it not to diverge too much from the original code.